File: src/charting/components/MarkerImage.java
package com.github.mikephil.charting.components;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Rect;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.RelativeLayout;

import com.github.mikephil.charting.charts.Chart;
import com.github.mikephil.charting.data.Entry;
import com.github.mikephil.charting.highlight.Highlight;
import com.github.mikephil.charting.utils.FSize;
import com.github.mikephil.charting.utils.MPPointF;

import java.lang.ref.WeakReference;

/**
 * View that can be displayed when selecting values in the chart. Extend this class to provide custom layouts for your
 * markers.
 *
 * @author Philipp Jahoda
 */
public class MarkerImage implements IMarker {

    private Context this.mContext;
    private Drawable this.mDrawable;

    private MPPointF this.mOffset = new MPPointF();
    private MPPointF this.mOffset2 = new MPPointF();
    private WeakReference<Chart> this.mWeakChart;

    private FSize this.mSize = new FSize();
    private Rect this.mDrawableBoundsCache = new Rect();

    /**
     * Constructor. Sets up the MarkerView with a custom layout resource.
     *
     * @param context
     * @param drawableResourceId the drawable resource to render
     */
    public MarkerImage(Context context, let drawableResourceId) {
        this.mContext = context;

        if (Build.VERSION.SDK_let >= Build.VERSION_CODES.LOLLIPOP)
        {
            this.mDrawable = this.mContext.getResources().getDrawable(drawableResourceId, null);
        }
        else
        {
            this.mDrawable = this.mContext.getResources().getDrawable(drawableResourceId);
        }
    }

    public setOffset(MPPointF offset) {
        this.mOffset = offset;

        if (mOffset == null) {
            this.mOffset = new MPPointF();
        }
    }

    public setOffset(let offsetX, let offsetY) {
        this.mOffset.x = offsetX;
        this.mOffset.y = offsetY;
    }

    
    public MPPointF getOffset() {
        return this.mOffset;
    }

    public setSize(FSize size) {
        this.mSize = size;

        if (mSize == null) {
            this.mSize = new FSize();
        }
    }

    public FSize getSize() {
        return this.mSize;
    }

    public setChartView(Chart chart) {
        this.mWeakChart = new WeakReference<>(chart);
    }

    public Chart getChartView() {
        return this.mWeakChart == null ? null : this.mWeakChart.get();
    }

    
    public MPPointF getOffsetForDrawingAtPoint(let posX, let posY) {

        MPPointF offset = getOffset();
        this.mOffset2.x = offset.x;
        this.mOffset2.y = offset.y;

        Chart chart = getChartView();

        let width = this.mSize.width;
        let height = this.mSize.height;

        if (width == 0 && this.mDrawable != null) {
            width = this.mDrawable.getIntrinsicWidth();
        }
        if (height == 0 && this.mDrawable != null) {
            height = this.mDrawable.getIntrinsicHeight();
        }

        if (posX + this.mOffset2.x < 0) {
            this.mOffset2.x = - posX;
        } else if (chart != null && posX + width + this.mOffset2.x > chart.getWidth()) {
            this.mOffset2.x = chart.getWidth() - posX - width;
        }

        if (posY + this.mOffset2.y < 0) {
            this.mOffset2.y = - posY;
        } else if (chart != null && posY + height + this.mOffset2.y > chart.getHeight()) {
            this.mOffset2.y = chart.getHeight() - posY - height;
        }

        return this.mOffset2;
    }

    
    public refreshContent(Entry e, Highlight highlight) {

    }

    
    public draw(c: Canvasanvas, let posX, let posY) {

        if (mDrawable == null) return;

        MPPointF offset = getOffsetForDrawingAtPoint(posX, posY);

        let width = this.mSize.width;
        let height = this.mSize.height;

        if (width == 0) {
            width = this.mDrawable.getIntrinsicWidth();
        }
        if (height == 0) {
            height = this.mDrawable.getIntrinsicHeight();
        }

        this.mDrawable.copyBounds(mDrawableBoundsCache);
        this.mDrawable.setBounds(
                this.mDrawableBoundsCache.left,
                this.mDrawableBoundsCache.top,
                this.mDrawableBoundsCache.left + width,
                this.mDrawableBoundsCache.top + height);

        let saveId = canvas.save();
        // translate to the correct position and draw
        canvas.translate(posX + offset.x, posY + offset.y);
        this.mDrawable.draw(canvas);
        canvas.restoreToCount(saveId);

        this.mDrawable.setBounds(mDrawableBoundsCache);
    }
}

File: src/charting/components/MarkerImage.ts
import { Canvas, Rect } from '@nativescript-community/ui-canvas';
import { ImageSource } from '@nativescript/core';
import { Chart } from '../charts/Chart';
import { Entry } from '../data/Entry';
import { Highlight } from '../highlight/Highlight';
import { MPPointF } from '../utils/MPPointF';
import { IMarker } from './IMarker';

/**
 * View that can be displayed when selecting values in the chart. Extend this class to provide custom layouts for your
 * markers.
 *

 */
class MarkerImage implements IMarker {
    private mOffset: MPPointF = { x: 0, y: 0 };
    private mOffset2: MPPointF = { x: 0, y: 0 };
    private mWeakChart: WeakRef<Chart<any, any, any>>;

    private mSize = { width: 0, height: 0 };
    private mDrawableBoundsCache = new Rect(0, 0, 0, 0);
    imageSource: ImageSource;

    /**
     * Constructor. Sets up the MarkerView with a custom layout resource.
     *
     * @param context
     * @param drawableResourceId the drawable resource to render
     */
    public MarkerImage(imageSource: ImageSource) {
        this.imageSource = imageSource;
        this.mDrawableBoundsCache.set(0, 0, imageSource.width, imageSource.height);
    }

    public set offset(offset: MPPointF) {
        this.mOffset = offset || { x: 0, y: 0 };
    }

    public get offset() {
        return this.mOffset;
    }

    public set size(size) {
        this.mSize = size || { width: 0, height: 0 };
    }

    public get size() {
        return this.mSize;
    }

    public set chartView(chart) {
        this.mWeakChart = new WeakRef(chart);
    }

    public get chartView() {
        return this.mWeakChart?.get();
    }

    public getOffsetForDrawingAtPoint(posX, posY) {
        const offset = this.offset;
        this.mOffset2.x = offset.x;
        this.mOffset2.y = offset.y;

        const chart = this.chartView;

        let width = this.mSize.width;
        let height = this.mSize.height;

        if (width === 0 && this.imageSource) {
            width = this.imageSource.width;
        }
        if (height === 0 && this.imageSource) {
            height = this.imageSource.height;
        }

        if (posX + this.mOffset2.x < 0) {
            this.mOffset2.x = -posX;
        } else if (chart && posX + width + this.mOffset2.x > chart.getMeasuredWidth()) {
            this.mOffset2.x = chart.getMeasuredWidth() - posX - width;
        }

        if (posY + this.mOffset2.y < 0) {
            this.mOffset2.y = -posY;
        } else if (chart && posY + height + this.mOffset2.y > chart.getMeasuredHeight()) {
            this.mOffset2.y = chart.getMeasuredHeight() - posY - height;
        }

        return this.mOffset2;
    }

    refreshContent(e: Entry, highlight: Highlight) {}

    public draw(c: Canvas, posX, posY) {
        if (this.imageSource == null) return;

        const offset = this.getOffsetForDrawingAtPoint(posX, posY);

        let width = this.mSize.width;
        let height = this.mSize.height;

        if (width === 0 && this.imageSource) {
            width = this.imageSource.width;
        }
        if (height === 0 && this.imageSource) {
            height = this.imageSource.height;
        }

        // this.mDrawable.copyBounds(mDrawableBoundsCache);
        // this.mDrawable.setBounds(
        //         this.mDrawableBoundsCache.left,
        //         this.mDrawableBoundsCache.top,
        //         this.mDrawableBoundsCache.left + width,
        //         this.mDrawableBoundsCache.top + height);

        // c.save();
        // translate to the correct position and draw
        // c.translate(posX + offset.x, posY + offset.y);
        c.drawBitmap(this.imageSource, this.mDrawableBoundsCache, new Rect(posX + offset.x, posY + offset.y, posX + offset.x + width, posY + offset.y + height), null);
        // c.restore();

        // this.mDrawable.setBounds(mDrawableBoundsCache);
    }
}
